[FEAT] 멘토링 리스트 조회 기능 구현

dye0p opened the pull request:
## Issue Number
closed #15

## As-Is
<!-- 문제 상황 정의 -->
멘토링 조회 api 필요

## To-Be
<!-- 변경 사항 -->
- 멘토링 조회 api 구현
 - 단건 조회
 - 전체 조회
 - 카테고리 필터링 조회

## Check List
- [x] 테스트가 전부 통과되었나요?
- [x] 모든 commit이 push 되었나요?
- [x] merge할 branch를 확인했나요?
- [x] Assignee를 지정했나요?
- [x] Label을 지정했나요?
- [x] 닫을 이슈 번호를 지정했나요?


## (Optional) Additional Description

jye326 committed:
build: 테스트 DB 환경 세팅 #15

Co-authored-by: dye0p <s01054830455@gmail.com>

jye326 committed:
refactor: 사진 테이블 칼럼명 변경 #15

- relationType -> imageType

Co-authored-by: dye0p <s01054830455@gmail.com>

jye326 committed:
feat: 전체 멘토링 조회 API 구현 #15

Co-authored-by: dye0p <s01054830455@gmail.com>

jye326 committed:
test: 전체 멘토링 조회 API 테스트 작성 #15

Co-authored-by: dye0p <s01054830455@gmail.com>

jye326 committed:
feat: 전체 멘토링 조회 API 카테고리 필터링 기능 추가 #15

- 아무 카테고리도 없으면 전체 조회

Co-authored-by: dye0p <s01054830455@gmail.com>

jye326 committed:
test: 전체 멘토링 조회 API 카테고리 필터링 기능 테스트 추가 #15

Co-authored-by: dye0p <s01054830455@gmail.com>

dye0p committed:
build: docker-compose 환경 세팅 #15

Co-authored-by: jye326 <juyongeun@gmail.com>

dye0p committed:
feat: 단일 멘토링 조회 API 구현 #15

Co-authored-by: jye326 <juyongeun@gmail.com>

dye0p committed:
test: 단일 멘토링 조회 테스트 작성 #15

Co-authored-by: jye326 <juyongeun@gmail.com>

ppparkta commented on backend/fittoring/src/main/java/fittoring/mentoring/business/model/Mentoring.java:
```diff
@@ -7,10 +7,10 @@
 import lombok.NoArgsConstructor;
 
-@AllArgsConstructor(access = AccessLevel.PRIVATE)
+@Getter
```
어떤 엔티티엔 필드에만 getter가 있고 어떤 엔티티는 최상단에 getter가 있네요!

모든 필드에 대해 getter를 쓰는 경우 클래스 레벨에 어노테이션을 쓰신 것은 이해되지만 최상단에 getter를 붙이지 않기로 논의했던 것 같아요.

그러나 개발이 진행되면서 오히려 필드에 붙은 getter를 지우고 최상단으로 올리는 등의 작업이 불편할 것 같기도 합니다😅

이 부분은 다시 얘기해보면 좋겠습니다.

dye0p replied:
엔티티 사용시 getter 사용이 잦아질 것 같은데. 이 부분을 고려해서 다시 얘기해봐요!

ppparkta commented on backend/fittoring/src/main/java/fittoring/mentoring/business/repository/ImageRepository.java:
```diff
@@ -1,9 +1,13 @@
 @Repository
 public interface ImageRepository extends ListCrudRepository<Image, Long> {
+
+    Optional<Image> findByImageTypeAndRelationId(ImageType imageType, Long id);
```
메서드명은 수정되지 않은 것 같아요

ppparkta replied:
이 부분 제가 착각했네요! 죄송합니다😢

ppparkta commented on backend/fittoring/src/main/resources/application.yml:
```diff
@@ -1,6 +1,6 @@
 spring:
   datasource:
-    url: jdbc:mysql://db:3306/fittoring?serverTimezone=Asia/Seoul&characterEncoding=UTF-8
+    url: jdbc:mysql://db:3306/fittoring
```
DB+서버 시간 불일치 문제로 인해 서버 시간 붙여주세요!
```
url: jdbc:mysql://db:3306/fittoring?serverTimezone=Asia/Seoul&characterEncoding=UTF-8
```

https://fftl.tistory.com/16

ppparkta commented on backend/fittoring/src/main/java/fittoring/mentoring/business/model/Image.java:
```diff
@@ -9,24 +9,33 @@
     @Enumerated(EnumType.STRING)
     @Column(nullable = false)
-    private ImageType relationType;
+    private ImageType imageType;
```
👍

ppparkta commented on backend/fittoring/src/test/resources/application.yml:
```diff
@@ -1,6 +1,6 @@
 spring:
   datasource:
-    url: jdbc:mysql://db-test:3306/fittoring-test?serverTimezone=Asia/Seoul&characterEncoding=UTF-8
+    url: jdbc:mysql://db-test:13307/fittoring-test?serverTimezone=Asia/Seoul&characterEncoding=UTF-8
```
이 부분도 3306으로 변경해주세요.

dye0p replied:
테스트 db 사용시 포트번호 13307 로 사용하는 것으로 알고있었는데, 아니었나요?

ppparkta replied:
맞습니다 로컬에서 작업하실 땐 13307로 사용해주시면 됩니다👍

도커 컨테이너가 prodDb, testDb 각각 따로 올라가서 로컬에서는 13307 포트를 사용하지만 도커 네트워크 상에서는 포트 매핑이 필요없으므로 기본 포트인 3306으로 연결해야 합니다. 그래서 올려주실 땐 3306으로 연결부탁드려용

dye0p replied:
이해했습니다! 감사합니다 :)

ppparkta commented on backend/fittoring/src/test/java/fittoring/mentoring/business/service/MentoringServiceTest.java:
👍

kimdevv commented on backend/fittoring/src/main/java/fittoring/mentoring/business/model/Category.java:
```diff
@@ -18,6 +19,11 @@ public class Category {
 
+    @Getter
     @Column(nullable = false, unique = true)
     private String title;
```
`@Getter`을 필드에도 달 수가 있었군요 👍

kimdevv commented on backend/fittoring/src/main/java/fittoring/mentoring/business/model/Mentoring.java:
```diff
@@ -36,26 +36,13 @@ public class Mentoring {
     @Column(nullable = false)
     private String introduction;
 
-    public Mentoring(
-            String mentorName,
-            String mentorPhone,
-            int price,
-            Integer career,
-            String content,
-            String introduction
-    ) {
-        this(null, mentorName, mentorPhone, price, career, content, introduction);
-    }
```
이런 식으로 `@AllArgsConstructor`을 사용해서 전체 필드를 가지는 생성자를 만들어 두고, 특정 매개변수를 가져야 하는 생성자가 생길 때마다 이 어노테이션으로 만든 생성자를 돌려 사용하면 재사용성을 높일 수 있지 않을까 생각합니다! 그렇지 않으면 생성자가 생길 때마다 `this.x= x`와 같은 똑같은 코드가 작성되어야 하니까요! 특정 필드에 `null`이 들어간다는 것도 명시할 수 있다는 부가 장점도 있을 것 같네요!

dye0p replied:
좋은 관점이네요. 저의 의견은 조금 다른데요. @AllArgsConstructor를 사용하는 목적은 **전체 필드를 초기화 하는 생성자를 자동으로 생성하는 것** 입니다. 그 말은, **생성자 초기화를 편리하게 하기 위해서**라고 생각합니다. 가령 **DTO의 생성자 초기화 정도가 생각나네요.(물론 DTO를 record로 생성한다면 사용할 필요가 없어집니다.)

어노테이션을 사용해서 모든 필드를 받는 생성자를 열고, 옵션으로 private 처리를 하여 중복되는 생성자를 재사용 한다는 관점은 개발 편의성을 높일 수는 있겠지만, 본래의 목적과 다른 방식으로 사용된다는 점과, 다른 개발자가 이 코드를 봤을 때, 혼란이 올 수도 있을 것 같습니다. 

명시적 생성자 사용으로도 말씀하신 null을 명시적으로 처리하여 의도를 드러낼 수 있을 것 같아요!
``` java
    public Mentoring(String mentorName, String mentorPhone, int price, Integer career, String content,
                     String introduction) {
        this.id = null; //null 명시
        this.mentorName = mentorName;
        this.mentorPhone = mentorPhone;
        this.price = price;
        this.career = career;
        this.content = content;
        this.introduction = introduction;
    }
```

만약, 매개변수가 각각 다른 생성자가 여러개 생긴다면, 그땐, 빌더를 고려해볼 수 있을 것 같아요.
빌더 패턴은 생성자의 매개변수와 개수와, 순서의 제약 없이 하나의 생성자로 객체를 생성할 수 있습니다! 
이 부분은 추후에 논의 해봐도 될 것 같아요!

ppparkta replied:
@kimdevv 사실 어제 가위바위보로 결정하기로 했는데 제가 져서 이 부분 삭제됐어요. 죄송함니다.

kimdevv replied:
❔

jye326 replied:
김데브씨 의견을 듣고 저도 다시 한번 생각해 봤는데요.
- 처음에는 ID가 없는 생성자를 만들기 위해 전체 생성자를 여는 것이 부담스럽게 느껴졌는데
- 이미 `Lombok`도 쓰고 있겠다, `@AllArgsConstructor`에 `PRRIVATE`을 걸어서 사용하는 게 생성자 부분 가독성이 높아질 것 같아요! 
다시 한번 논의를 해보면 좋을 것 같아요!

kimdevv commented on backend/fittoring/src/main/java/fittoring/mentoring/business/repository/ImageRepository.java:
```diff
@@ -1,9 +1,13 @@
 @Repository
 public interface ImageRepository extends ListCrudRepository<Image, Long> {
+
+    Optional<Image> findByImageTypeAndRelationId(ImageType imageType, Long id);
```
변수 이름을 `id` 대신 `relationId`로 하는 것은 어떨까요?

dye0p replied:
그렇게 하면 명시적으로 보이겠네요👍

jye326 replied:
그렇군요! 지금 상태로는 ImageId와 헷갈릴 것 같아요! 수정하겠습니다!

kimdevv commented on backend/fittoring/src/main/java/fittoring/mentoring/business/service/MentoringService.java:
```diff
@@ -0,0 +1,103 @@
+    }
+
+    private List<String> getCategoryTitlesByMentoring(List<CategoryMentoring> categoriesByMentoring) {
+        List<String> categoryTitles = new ArrayList<>();
+        for (CategoryMentoring categoryMentoring : categoriesByMentoring) {
+            categoryTitles.add(categoryMentoring.getCategoryTitle());
+        }
+        return categoryTitles;
```
stream 대신 for문을 사용하신 이유가 궁금합니다!

dye0p replied:
특별한 이유는 없었습니다. 스트림 형식이 익숙치 않아서 우선 for문으로 작성했습니다!

jye326 replied:
일단 빠르게 기능을 구현하는데 집중하자는 취지에서 익숙한 for문을 사용했습니다.
가독성을 위해 stream으로 변환할 수 있습니다.

ppparkta replied:
저는 단순연산에는 for문의 가독성이 나쁘지 않다고 생각해요.

kimdevv commented on backend/fittoring/src/main/java/fittoring/mentoring/business/service/MentoringService.java:
```diff
@@ -0,0 +1,103 @@
+        validateExistCategoryTitle(categoryTitle3);
+    }
+
+    private void validateExistCategoryTitle(String categoryTitle1) {
```
파라미터 이름을 `categoryTitle`으로 변경하는 게 어떨까요?

dye0p replied:
꼼꼼하게 확인하지 못했네요..!

jye326 replied:
놓친 부분이 있었네요..!

kimdevv commented on backend/fittoring/src/main/resources/application.yml:
```diff
@@ -1,6 +1,6 @@
 spring:
   datasource:
-    url: jdbc:mysql://db:3306/fittoring?serverTimezone=Asia/Seoul&characterEncoding=UTF-8
```
타임존 관련 설정도 추가하면 좋을 것 같아요!

dye0p replied:
추가했습니다!

kimdevv commented on backend/fittoring/src/test/java/fittoring/mentoring/business/service/MentoringServiceTest.java:
```diff
@@ -0,0 +1,303 @@
+                    .containsExactly(response, response2);
+        }
+
+        @DisplayName("검색 필터가 하나인 경우 ")
```
`@DisplayName`을 작성할 때 **"~다."** 형태로 작성하는 컨벤션이 있어요!

dye0p replied:
꼼꼼하게 확인하지 못했네요. 수정하겠습니다!

jye326 replied:
놓친 부분이 있었네요..!

kimdevv commented on backend/fittoring/src/test/java/fittoring/mentoring/business/service/MentoringServiceTest.java:
```diff
@@ -0,0 +1,303 @@
+
+            // when
+            // then
+            MentoringResponse response = MentoringResponse.from(mentoring1,
+                    List.of(categoryMentoring1.getCategoryTitle()),
+                    image1);
+            MentoringResponse response2 = MentoringResponse.from(mentoring2,
+                    List.of(categoryMentoring2.getCategoryTitle()));
```
저 같은 경우는 이 부분을 given에다가 쓰고 변수명을 `expected`로 작성하는 편인데요...! 이처럼  then에는 결과를 검증하는 코드만 쓰는 방식은 어떤가요??

dye0p replied:
then에 해당 로직을 추가한 이유는 response, response2 인스턴스가 **검증에 쓰일 값들** 이라는 의미를 표현하기 위해서 였습니다. then으로 결과의 의도를 표현하고 있으니 말씀해주신 것 처럼 given절에 올려도 의미 표현에 문제가 없어 보입니다!

jye326 replied:
given 절에 올렸습니다!

kimdevv reviewed (approved):
고생 많으셨습니다! 어프루브할게요~

ppparkta commented on frontend/.storybook/preview.ts:
프론트 코드 커밋에 같이 올라가도 되나요??

dye0p replied:
아, 실수로 같이 커밋한 것 같은데 어떻게 되는지 잘 모르겠네요..

jye326 replied:
백엔드 커밋만 남겼습니다!

ppparkta commented on backend/fittoring/src/main/java/fittoring/mentoring/presentation/api/MentoringController.java:
```diff
@@ -0,0 +1,34 @@
+
+    private final MentoringService mentoringService;
+
+    @GetMapping("/mentorings")
+    public ResponseEntity<List<MentoringResponse>> getAllMentoring(
+            @RequestParam(required = false) String categoryTitle1,
+            @RequestParam(required = false) String categoryTitle2,
+            @RequestParam(required = false) String categoryTitle3
+    ) {
+        return ResponseEntity.ok()
+                .body(mentoringService.getAllMentoring(categoryTitle1, categoryTitle2, categoryTitle3));
+    }
```
어제 설명해주신 방법대로 잘 구현해주셨네요 좋습니다!! 💪

추가적으로 지금 코드는 카테고리 필터링이 최대 세개만 가능한데, 검색할 때는 카테고리 갯수 제한이 따로 없는 것으로 알고있어요. 

이 부분을 해소하기 위해서는 List<String> 형태로 파라미터를 받고, 기존 쿼리를 변경하거나 Querydsl 동적쿼리로 개선할 수 있을 것 같아요!

dye0p replied:
2차 데모까지 정책이 "카테고리 검색시 갯수 제한이 없다" 였나요? 그게 맞다면 제가 잘못 이해 하고 있었던 것 같네요. 최대 3개까지 카테고리 필터링이 가능한줄 알았습니다. 

카테고리 필터링의 개수 제한이 없다면 말씀해주신 것 처럼 List 형태로 값을 받아야 할 것 같습니다. 👍

ppparkta commented on backend/fittoring/src/main/java/fittoring/mentoring/business/repository/MentoringRepository.java:
```diff
@@ -1,10 +1,33 @@
+                      COALESCE(:title2, '__NULL__'),
+                      COALESCE(:title3, '__NULL__')
+                  )
+             GROUP BY m.id
+             HAVING COUNT (DISTINCT cm.category.title) =
```
이 부분 어떻게 필터링되고 있나요?

카테고리 필터링은 and 연산이 아니라 or 연산으로 필터링 되어야 하는데, 지금은 having 절에서 요청한 카테고리 개수와 매칭된 개수를 비교하고 있어 and 연산처럼 구현된 것 같아요.

ppparkta replied:
앞에서 말씀드린 파라미터를 리스트로 받는 형태로 or연산 쿼리짜봤어요!

```java
@Query("""
    select distinct m
    from categoryMentoring cm
    join cm.mentoring m
    where cm.category.title in :titles
""")
List<Mentoring> 메서드명(@Param("titles") List<String> titles);
```

dye0p replied:
> 이 부분 어떻게 필터링되고 있나요?
카테고리 필터링은 and 연산이 아니라 or 연산으로 필터링 되어야 하는데, 지금은 having 절에서 요청한 카테고리 개수와 매칭된 개수를 비교하고 있어 and 연산처럼 구현된 것 같아요.

이 부분은  @jye326 과 다시 검토해보겠습니다. SQL 지식이 부족해서 잘못 짜였을 수도 있을 것 같아요! 지적 감사합니다👍

dye0p replied:
> > 이 부분 어떻게 필터링되고 있나요?
> > 카테고리 필터링은 and 연산이 아니라 or 연산으로 필터링 되어야 하는데, 지금은 having 절에서 요청한 카테고리 개수와 매칭된 개수를 비교하고 있어 and 연산처럼 구현된 것 같아요.
> 
> 이 부분은 @jye326 과 다시 검토해보겠습니다. SQL 지식이 부족해서 잘못 짜였을 수도 있을 것 같아요! 지적 감사합니다👍

이 부분은 정책 이해 관계의 차이가 있었습니다. 팀 회의를 통해 우선 카테고리 필터링을 AND 조건으로 진행하기로 결정했습니다!

ppparkta reviewed (commented):
코드 전부 확인했어요! 고생많으셨습니다 💪
코멘트 몇 개 남겼는데 확인 부탁드려요!

ppparkta committed:
Merge branch 'develop' into feature/15-read-mentoring-list

dye0p committed:
refactor: findByImageTypeAndRelationId()매개변수 네이밍 변경

- id -> relationId

Co-authored-by: jye326 <juyongeun@gmail.com>

dye0p committed:
refactor: validateExistCategoryTitle() 매개변수 네이밍 변경

- categoryTitle1 -> categoryTitle

Co-authored-by: jye326 <juyongeun@gmail.com>

dye0p committed:
fix: 카테고리 기능 버그 수정

- 필요 카테고리를 충족하면서 별도의 카테고리를 포함하는 멘토링이 조회가 안되던 문제를 수정

Co-authored-by: jye326 <juyongeun@gmail.com>

dye0p committed:
refactor: 카테고리_멘토링 클래스 필드 순서 변경

- 기존, 클래스네이밍과 반대로 되어있던 필드 순서를 클래스 네이밍의 순서에 맞게 변경

Co-authored-by: jye326 <juyongeun@gmail.com>

dye0p committed:
test: MentoringService 테스트 수정

- 카테고리 필터링 조건을 만족하는 모든 멘토링을 조회하는 테스트 추가

Co-authored-by: jye326 <juyongeun@gmail.com>

dye0p committed:
refactor: MentoringService 클래스 리팩터링

- foreach 문 stream 으로 변경
- 변수명 명확하게 변경

Co-authored-by: jye326 <juyongeun@gmail.com>

kimdevv commented on backend/fittoring/src/main/java/fittoring/mentoring/business/model/CategoryMentoring.java:
```diff
@@ -21,9 +23,18 @@ public class CategoryMentoring {
+        return category.getTitle();
+    }
+
+    public CategoryMentoring(final Category category, final Mentoring mentoring) {
```
p2: 저희 `final`은 붙이지 않기로 했던 것 같아요!

kimdevv commented on backend/fittoring/src/main/java/fittoring/mentoring/business/repository/MentoringRepository.java:
```diff
@@ -1,10 +1,33 @@
 public interface MentoringRepository extends ListCrudRepository<Mentoring, Long> {
 
+    @Query("""
+            SELECT m
+              FROM CategoryMentoring cm
+              JOIN cm.mentoring m
+             WHERE cm.category.title IN (
+                      COALESCE(:title1, '__NULL__'),
+                      COALESCE(:title2, '__NULL__'),
+                      COALESCE(:title3, '__NULL__')
+                  )
+             GROUP BY m.id
+             HAVING COUNT (DISTINCT cm.category.title) >=
+                (CASE WHEN :title1 IS NOT NULL THEN 1 ELSE 0 END +
+                 CASE WHEN :title2 IS NOT NULL THEN 1 ELSE 0 END +
+                 CASE WHEN :title3 IS NOT NULL THEN 1 ELSE 0 END)
```
p3: 파라미터로 들어온 title의 개수 이상을 가진 cm을 먼저 필터링한 후에, 그 중에서 파라미터의 title을 모두 포함한 cm과 m을 조인해서 m을 선택하는 걸로 이해했는데 맞을까요??

jye326 replied:
where절이 먼저 적용된 후 title 갯수 이상을 가진 mentoring을 조회합니다.

kimdevv commented on backend/fittoring/src/main/java/fittoring/mentoring/business/service/MentoringService.java:
```diff
@@ -0,0 +1,104 @@
+    private final CategoryMentoringRepository categoryMentoringRepository;
+    private final ImageRepository imageRepository;
+
+    public List<MentoringResponse> getAllMentoring(
```
p2: 모든 멘토링을 가져올 수도 있지만 필터링해서 조회할 수도 있기 때문에 메서드명을 수정해야할 것 같습니다! `findMentorings` 정도...?

dye0p replied:
무조건 존재한다의 의미인 get 보단, 존재할 수도 있다는 의미의 find를 사용하도록 변경했습니다.
추가로, 필터링에 따라서 모든 멘토링을 조회하지 않을 수도 있기 때문에 All이라는 단어를 제거하였습니다.

kimdevv commented on backend/fittoring/src/main/java/fittoring/mentoring/presentation/api/MentoringController.java:
```diff
@@ -0,0 +1,34 @@
+    @GetMapping("/mentorings/{mentoringId}")
+    public ResponseEntity<MentoringResponse> getMentoring(@PathVariable(name = "mentoringId") Long id) {
+        MentoringResponse response = mentoringService.getMentoring(id);
+        return ResponseEntity.ok(response);
```
p2:이 부분 통일하는 거에 대해서도 얘기해보면 좋을 것 같아요!

kimdevv reviewed (approved):
고생하셨습니다 저는 approve할게요~

ppparkta commented on backend/fittoring/src/main/java/fittoring/mentoring/business/model/Image.java:
```diff
@@ -9,24 +9,33 @@
+        this.url = url;
+        this.imageType = imageType;
+        this.relationId = relationId;
+    }
```
p1: 이 부분 생성자도 통일되면 좋을 것 같아요!

ppparkta commented on backend/fittoring/src/main/java/fittoring/mentoring/presentation/api/MentoringController.java:
```diff
@@ -0,0 +1,34 @@
+    }
+
+    @GetMapping("/mentorings/{mentoringId}")
+    public ResponseEntity<MentoringResponse> getMentoring(@PathVariable(name = "mentoringId") Long id) {
```
p1: `@PathVariable(name="mentoringId")`, `@PathVariable("mentoringId")`  
이전에 merge된 코드는 후자로 작성되었어요. 일관성 유지 위해 한쪽으로 통일하면 좋을 것 같습니다.

dye0p replied:
기능의 차이가 없는 것을 확인했습니다. 후자가 가독성이 더 좋겠네요! 변경하겠습니다.

ppparkta commented on backend/fittoring/src/test/java/fittoring/mentoring/business/service/MentoringServiceTest.java:
```diff
@@ -0,0 +1,288 @@
+
+@ActiveProfiles("test")
+@AutoConfigureTestDatabase(replace = Replace.NONE)
+@Import(MentoringService.class)
```
p2: 생성하고 있는 엔티티가 많은 것 같은데 DbCleaner 필요하지 않을까요?

ppparkta commented on backend/fittoring/src/main/java/fittoring/mentoring/business/repository/MentoringRepository.java:
```diff
@@ -1,10 +1,33 @@
 @Repository
 public interface MentoringRepository extends ListCrudRepository<Mentoring, Long> {
 
+    @Query("""
+            SELECT m
+              FROM CategoryMentoring cm
+              JOIN cm.mentoring m
+             WHERE cm.category.title IN (
+                      COALESCE(:title1, '__NULL__'),
+                      COALESCE(:title2, '__NULL__'),
+                      COALESCE(:title3, '__NULL__')
+                  )
+             GROUP BY m.id
+             HAVING COUNT (DISTINCT cm.category.title) >=
+                (CASE WHEN :title1 IS NOT NULL THEN 1 ELSE 0 END +
+                 CASE WHEN :title2 IS NOT NULL THEN 1 ELSE 0 END +
+                 CASE WHEN :title3 IS NOT NULL THEN 1 ELSE 0 END)
```
p3: in 서브쿼리에 `COALESCE(anyTitle, '__NULL__')`이 필요한 이유는 타이틀이 null일 경우 `__NULL__`이라는 DB에 들어있지 않은 임의의 값을 대신 조회해서 조회결과에 포함되지 않게 하기 위해서,
group by/having이 필요한 이유는 하나의 멘토링에 포함된 카테고리의 개수를 비교하기 위해서 맞을까요?

jye326 replied:
맞습니다!

ppparkta reviewed (approved):
토의가 필요한 내용 코멘트로 남겨두었습니다. 
컨벤션 부분만 맞춰보고 바로 merge해도 될 것 같아 approve하겠습니다. 고생하셨어요!

dye0p committed:
refactor: 이미지 엔티티 생성자 리팩터링

- 생성자 방식 팀 컨벤션에 맞게 변경

Co-authored-by: jye326 <juyongeun@gmail.com>

dye0p committed:
refactor: 멘토링 조회 메서드 네이밍 명확하게 변경

- 무조건 존재한다의 의미인 'get'보단, 존재할 수도 있다는 의미의 'find'로 변경
-  필터링에 따라서 모든 멘토링을 조회하지 않을 수도 있기 때문에 'All'을 제거함

Co-authored-by: jye326 <juyongeun@gmail.com>

dye0p committed:
refactor: @PathVariable name 속성 단축 표현으로 변경

- 더 나은 가독성을 위해서 명시적인 name속성 대신 단축 표현으로 변경

Co-authored-by: jye326 <juyongeun@gmail.com>

dye0p committed:
test: 테스트 DBCleaner 추가

Co-authored-by: jye326 <juyongeun@gmail.com>

dye0p committed:
refactor: 생성자 final 키워드 제거

Co-authored-by: jye326 <juyongeun@gmail.com>
